### PR TITLE
Add metric to count search responses

### PR DESCRIPTION
**Pull request by jdconrad**
This adds a metric to count the total number of search responses generated by the search action and the scroll search action. This uses the key `es.search_response.response_count.total`. Each response also has an attribute to specify whether the response was a success, a partial failure, or a failure.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/action/search/SearchResponseMetrics.java [hunk 1]
```diff
@@ -8,31 +8,55 @@
...
         this(
             meterRegistry.registerLongHistogram(
                 TOOK_DURATION_TOTAL_HISTOGRAM_NAME,
                 "The SearchResponse.took durations in milliseconds, expressed as a histogram",
                 "millis"
+            ),
+            meterRegistry.registerLongCounter(
+                RESPONSE_COUNT_TOTAL_COUNTER_NAME,
+                "The cumulative total of search responses with an attribute to describe"
+                    + "success, partial failure, or failure, expressed as a counter and an attribute",
```
Comment (quux00) [inline server/src/main/java/org/elasticsearch/rest/action/search/SearchResponseMetrics.java:43]: nit: I think you need a space after "describe" or before "success"

Comment (quux00) [reply server/src/main/java/org/elasticsearch/rest/action/search/SearchResponseMetrics.java:43]: Asking to learn: what does "expressed as a counter and an attribute" mean? How is it both?

File: server/src/main/java/org/elasticsearch/rest/action/search/SearchResponseMetrics.java [hunk 2]
```diff
@@ -8,31 +8,55 @@
...
     }
 
     public long recordTookTime(long tookTime) {
         tookDurationTotalMillisHistogram.record(tookTime);
         return tookTime;
     }
+
+    public void incrementResponseCount(ResponseCountTotalStatus responseCountTotalStatus) {
+        responseCountTotalCounter.incrementBy(1L, Map.of(RESPONSE_COUNT_TOTAL_STATUS_ATTRIBUTE_NAME, responseCountTotalStatus));
+    }
```
Comment (JVerwolf) [inline server/src/main/java/org/elasticsearch/rest/action/search/SearchResponseMetrics.java:61]: This is what's causing the following build failure:

https://buildkite.com/elastic/elasticsearch-pull-request/builds/23506#01903d0a-f3e5-4f38-8025-68037928d91b
 
 ```
 Caused by: java.lang.IllegalArgumentException: attributes do not support value type of [org.elasticsearch.rest.action.search.SearchResponseMetrics.ResponseCountTotalStatus]
»       at org.elasticsearch.telemetry.apm@8.15.0-SNAPSHOT/org.elasticsearch.telemetry.apm.internal.metrics.OtelHelper.lambda$fromMap$0(OtelHelper.java:50)
```


You could change this to:

```
responseCountTotalCounter.incrementBy(1L, Map.of(RESPONSE_COUNT_TOTAL_STATUS_ATTRIBUTE_NAME, responseCountTotalStatus.name()));
}
```

Or,

```

    public enum ResponseCountTotalStatus {
        SUCCESS("success"),
        PARTIAL_FAILURE("partial_failure"),
        FAILURE("failure");
        private final String value;

        ResponseCountTotalStatus(String value) {
            this.value = value;
        }

        public String value() {
            return value
        }
    }
    ...
    
        responseCountTotalCounter.incrementBy(1L, Map.of(RESPONSE_COUNT_TOTAL_STATUS_ATTRIBUTE_NAME, responseCountTotalStatus.value()));

   ```

Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/rest/action/search/SearchResponseMetrics.java:61]: After chatting with @original-brownbear my concerns around the assert triggered by this error have been alleviated. My understanding is the assert is related to ensuring we don't except any exceptions to be triggered as part of `onResponse` except for unusual cases. I will fix this and then expect the tests will pass.

Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/rest/action/search/SearchResponseMetrics.java]: Changed to your suggestion of `value` but used `displayName` instead to not overload the term `value`.

**Comment by jdconrad** [comment]
@quux00 @JVerwolf Thanks for the initial review. I have responded to your feedback, and this is ready for another round.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 3]
```diff
@@ -292,24 +292,43 @@ public long buildTookInMillis() {
...
+                        // Deduplicate failures by exception message and index
+                        ShardOperationFailedException[] groupedFailures = ExceptionsHelper.groupBy(searchResponse.getShardFailures());
+                        for (ShardOperationFailedException f : groupedFailures) {
+                            boolean causeHas500Status = false;
+                            if (f.getCause() != null) {
+                                causeHas500Status = ExceptionsHelper.status(f.getCause()).getStatus() >= 500;
+                            }
+                            if ((f.status().getStatus() >= 500 || causeHas500Status)
+                                && ExceptionsHelper.isNodeOrShardUnavailableTypeException(f.getCause()) == false) {
+                                logger.warn("TransportSearchAction shard failure (partial results response)", f);
```
Comment (JVerwolf) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:312]: @quux00 Do you know why we only log partial failures when the response code is a 5xx?  I see that the metric that @jdconrad added follows the same example, so I was curious.

**Review by JVerwolf — APPROVED** [review]
LGTM!

Non-blocking: @quux00 Does https://github.com/jdconrad/elasticsearch/blob/c252b40ea63aa764c4a88e12d37f1e51af9af768/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java#L158 need to be instrumented as well?

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java (hunk 3 above)
Comment (quux00) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:312]: This was done in the context of serverless, where we don't log or alert on "user errors" (4xx) (other than 429s, but that doesn't apply here). The goal of the logging epic was to get a sense of how often 5xx "system" type errors are occurring when users get partial results.

**Comment by quux00** [comment]
> LGTM!
> 
> Non-blocking: @quux00 Does https://github.com/jdconrad/elasticsearch/blob/c252b40ea63aa764c4a88e12d37f1e51af9af768/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java#L158 need to be instrumented as well?

Yeah, probably. We could bring that to an es search foundations team meeting for discussion and follow on ticket.

**Pull request merged**